### PR TITLE
kubeadm: don't add proxy flag to apiservers below v1.7

### PR DESCRIPTION
**Pull request by mikedanese**

### CODE REVIEW

**Comment by mikedanese** [comment]
@k8s-bot kubeadm test this

File: cmd/kubeadm/app/master/manifests.go [hunk 1]
```diff
@@ -68,12 +69,17 @@ func WriteStaticPodManifests(cfg *kubeadmapi.MasterConfiguration) error {
 		volumeMounts = append(volumeMounts, pkiVolumeMount())
 	}
 
+	k8sVersion, err := version.ParseSemantic(cfg.KubernetesVersion)
```
Comment (pipejakob) [inline cmd/kubeadm/app/master/manifests.go]: I know on the command-line, we support strings like `latest` and `stable-1.6`. Are those guaranteed to be normalized to semver by this point?

File: cmd/kubeadm/app/master/selfhosted.go [hunk 2]
```diff
@@ -184,6 +185,8 @@ func waitForPodsWithLabel(client *clientset.Clientset, appLabel string, mustBeRu
 
 // Sources from bootkube templates.go
 func getAPIServerDS(cfg *kubeadmapi.MasterConfiguration, volumes []v1.Volume, volumeMounts []v1.VolumeMount) ext.DaemonSet {
+	//TODO: assert that this won't fail
```
Comment (pipejakob) [inline cmd/kubeadm/app/master/selfhosted.go]: Is this a TODO for this PR, or a future PR? Panicking is pretty hardcore.

File: cmd/kubeadm/app/master/manifests_test.go [hunk 3]
```diff
@@ -539,7 +538,7 @@ func TestGetAPIServerCommand(t *testing.T) {
 	}
 
 	for _, rt := range tests {
-		actual := getAPIServerCommand(rt.cfg, false)
+		actual := getAPIServerCommand(rt.cfg, false, version.MustParseSemantic("v1.6"))
```
Comment (pipejakob) [inline cmd/kubeadm/app/master/manifests_test.go]: How about a small test for 1.7 that asserts that the proxy-client flags get added?

File: cmd/kubeadm/app/master/manifests.go (hunk 1 above)
Comment (mikedanese) [inline cmd/kubeadm/app/master/manifests.go]: Yes they are.

**Comment by jellonek** [comment]
Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[cmd/kubeadm/app/master/manifests_test.go, line 541 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44002#-Kgpjsa0BJNvDw_6Crlk-r1-541:-KgsAnCLVbLA-SovnOV1:bb6mt9b) ([raw file](https://github.com/kubernetes/kubernetes/blob/9340aa2b6b296714de652b40a9109d335da9c0c2/cmd/kubeadm/app/master/manifests_test.go#L541)):*
<details><summary><i>Previously, pipejakob (Jakob) wrote…</i></summary><blockquote>

How about a small test for 1.7 that asserts that the proxy-client flags get added?
</blockquote></details>

:+1:

---

*[cmd/kubeadm/app/master/manifests_test.go, line 71 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44002#-KgsA6DN70KNA4SJSED3:-KgsA6DN70KNA4SJSED4:b-fac8vl) ([raw file](https://github.com/kubernetes/kubernetes/blob/e682d79e50a4c2c03390bf6ad05bd18761f5081a/cmd/kubeadm/app/master/manifests_test.go#L71)):*
> ```Go
> 		actual := WriteStaticPodManifests(rt.cfg)
> 		if (actual == nil) && rt.expectErr {
> 			t.Errorf("expected an error from WriteStaticPodManifests but got none")
> ```

This is plain string, so `t.Error` can be used instead of `t.Errorf`.

---

*[cmd/kubeadm/app/master/manifests_test.go, line 85 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44002#-KgsCKba24tcEggAb4ar:-KgsCKba24tcEggAb4as:b-be22o4) ([raw file](https://github.com/kubernetes/kubernetes/blob/e682d79e50a4c2c03390bf6ad05bd18761f5081a/cmd/kubeadm/app/master/manifests_test.go#L85)):*
> ```Go
> 		if rt.expectedAPIProbePort != 0 {
> 			path := fmt.Sprintf("%s/manifests/kube-apiserver.yaml", kubeadmapi.GlobalEnvParams.KubernetesDir)
> 			manifest, err := os.Open(path)
> ```

Why this is divided? Maybe you wanted to include path later in error message?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/44002)*

File: cmd/kubeadm/app/master/manifests.go [hunk 4]
```diff
@@ -323,9 +329,11 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, selfHosted bool) [
 		"requestheader-extra-headers-prefix": "X-Remote-Extra-",
 		"requestheader-client-ca-file":       path.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertName),
 		"requestheader-allowed-names":        "front-proxy-client",
+	}
+	if k8sVersion.AtLeast(version.MustParseSemantic("v1.7.0")) {
```
Comment (luxas) [inline cmd/kubeadm/app/master/manifests.go]: Will this work for v1.7.0-alpha.0 as well?
Can't you just say `k8sVersion.AtLeast("v1.7.0-alpha.1")` (and move that constant elsewhere)?

File: cmd/kubeadm/app/master/manifests_test.go [hunk 5]
```diff
@@ -539,7 +539,7 @@ func TestGetAPIServerCommand(t *testing.T) {
 	}
 
 	for _, rt := range tests {
-		actual := getAPIServerCommand(rt.cfg, false)
+		actual := getAPIServerCommand(rt.cfg, false, version.MustParseSemantic("v1.6.0"))
```
Comment (luxas) [inline cmd/kubeadm/app/master/manifests_test.go]: pls add a test to make sure that with v1.7.0-alpha.1 it works as expected

**Comment by mikedanese** [comment]
Fixed. PTAL.

File: cmd/kubeadm/app/master/manifests_test.go [hunk 6]
```diff
@@ -760,3 +800,18 @@ func TestGetExtraParameters(t *testing.T) {
...
+	versions := []string{
+		"v1.7.0-alpha.1",
+		"v1.7.0-beta.0",
+		"v1.7.0-rc.0",
+		"v1.7.0",
+		"v1.7.1",
+	}
+	for _, v := range versions {
+		if !version.MustParseSemantic(v).AtLeast(v170) {
+			t.Errorf("err")
```
Comment (pipejakob) [inline cmd/kubeadm/app/master/manifests_test.go]: Nitpick: could just be Error

**Review by pipejakob — APPROVED** [review]

**Pull request merged**